### ISSUE TITLE
[Discussion] Minimum register size platforms we support

### ISSUE THREAD

**Issue by dongcarl** [rust-bitcoin/rust-bitcoin#217]
What are people's thoughts on what the minimum register size platform we should support?

I am asking because there are a few `Copy` types in `rust-bitcoin` where their internal representation is very small (`u8` or `u32`). This means that it would be more efficient to pass by value.

This is inspired by: https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

**Comment by dpc** [comment]
Is the register size even relevant for optimization? All trivially copy-able primitives are probably better off  passed by value, from `u8` to `u128`, irrelevant of architecture. The bigger register-sized architectures, have easier time passing even the bigger values, but smaller-sized ones are more susceptible to have primitive micro-architecture and smaller (if any) caches, which makes the the cost of (de)referencing higher. On top of, since rust ABI is unspecified, the compiler has a freedom to ignore `&` and convert passing by value into passing by reference or the other way around anyway. So passing by value will be the easiest for the compiler to optimize anyway.

**Comment by dongcarl** [comment]
@dpc I'm talking more about our own types rather than primitives :relaxed:

**Comment by dpc** [comment]
@dongcarl No difference, as far as I can tell. All that matters is the size of the type/object.

**Comment by dongcarl** [comment]
@dpc Perhaps I'm misunderstanding you, you're not suggesting that all methods should take `self` instead of `&self` are you?

**Comment by dpc** [comment]
I guess we're in total misunderstanding. :D

I'm really making a point that types that are small are probably better off on any platform (confused by subject "Minimum register size platforms we support ").

> all methods should take `self` instead of `&self` are you

If the receiver (`Self`) is trivially `Copy` and small, then it's probably better to have it taken by `self`, for the same reasons as apply to any other function argument.

**Comment by dongcarl** [comment]
@dpc Right. I guess I'm asking: what is our definition of "small".

I know that there are real-world deployments of `rust-bitcoin` with `32-bit` architecture, so do we define "small" as <= 32-bit? (Which I'd be fine with honestly)

**Comment by dpc** [comment]
"Small" value? 64B is still "small" for me. As I said - compiler is free to do whatever it wants for passing value by copy, even pass them by reference if so it desires. So the difference is more in semantics, than perf. Unless dynamic dispatch is needed, by-copy is just easier to use.

I feel like you might have a prior C/C++ experience, and feel like values will be taking registers and there's going to be a lot of copying etc. I did had a lot of problem switching my mental model when I started working with Rust. But the same rules don't apply. ABI is undefined on purpose, monorphization, LTO, RVO are there by default and extensively relied on etc.

So for me any value is `Copy` if it doesn't require cloning or some other operations, and if the value is `Copy` I generally take it be `Self` everywhere and let compiler worry about optimizing calls (except for stuff like `extern "C"`, or maybe if it's reeeally big).
